src/tools/schema2base.ts

```diff
@@ -7,75 +7,108 @@
-
-    case 'schema-collection':
-      return `schema-collection:${descriptor.schema.type}`;
-
-    default:
-      return descriptor.kind;
+import {Dictionary} from '../runtime/hot.js';
+
+export abstract class Schema2Base {
+  constructor(readonly opts: minimist.ParsedArgs) {}
+
+  async call() {
+    fs.mkdirSync(this.opts.outdir, {recursive: true});
+    for (const file of this.opts._) {
+      if (fs.existsSync(file)) {
+        await this.processFile(file);
+      } else {
+        throw new Error(`File not found: ${file}`);
+      }
+    }
   }
-}
 
+  private async processFile(src: string) {
+    const outName = this.outputName(path.basename(src));
+    const outPath = path.join(this.opts.outdir, outName);
+    console.log(outPath);
+    if (this.opts.update && fs.existsSync(outPath) && fs.statSync(outPath).mtimeMs > fs.statSync(src).mtimeMs) {
+      return;
+    }
 
-export class Schema2Base {
-  private desc: string;
-  private fileNamer: (schemaName: string) => string;
-  private generate: (name: string, schema) => string;
-
-  constructor(description: string, fileNamer: (schemaName: string) => string, generator: (name: string, schema) => string) {
-    this.desc = description;
-    this.fileNamer = fileNamer;
-    this.generate = generator;
-  }
+    const contents = fs.readFileSync(src, 'utf-8');
+    const manifest = await Manifest.parse(contents);
 
+    // Collect declared schemas along with any inlined in particle connections.
+    const schemas: Dictionary<Schema> = {...manifest.schemas};
+    for (const particle of manifest.particles) {
+      for (const connection of particle.connections) {
+        const schema = connection.type.getEntitySchema();
+        const name = schema && schema.names && schema.names[0];
+        if (name && !(name in schemas)) {
+          schemas[name] = schema;
+        }
+      }
+    }
+    if (Object.keys(schemas).length === 0) {
+      console.warn(`No schemas found in '${src}'`);
+      return;
+    }
 
-  // TODO: handle schemas with multiple names and schemas with parents
-  // TODO: error handling
-  async processFile(file: string, outputPrefix: string = ''): Promise<void> {
-    const contents = fs.readFileSync(file, 'utf-8');
-    const manifest = await Manifest.parse(contents);
-    for (const schema of Object.values(manifest.schemas)) {
-      const outFile = this.fileNamer(schema.names[0]);
-      const contents = this.generate(schema.names[0], schema);
-      fs.writeFileSync(outputPrefix + outFile, contents);
+    const outFile = fs.openSync(outPath, 'w');
+    fs.writeSync(outFile, this.fileHeader(outName));
+    for (const [name, schema] of Object.entries(schemas)) {
+      fs.writeSync(outFile, this.entityClass(name, schema).replace(/ +\n/g, '\n'));
     }
+    fs.writeSync(outFile, this.fileFooter());
+    fs.closeSync(outFile);
   }
 
-  get scriptName() {
-    return path.basename(__filename).split('.')[0];
+  protected processSchema(schema: Schema, processField): number {
```

Comment:
Please fill in types for `processField`: 
`processField: (field: string, typeChar: string) => void`

Reply:
I like that this function is unified.